src/pathfinder/yapf/yapf_ship.cpp

```diff
@@ -388,6 +388,12 @@ class CYapfCostShipT {
 		uint8_t speed_frac = (GetEffectiveWaterClass(n.GetTile()) == WATER_CLASS_SEA) ? svi->ocean_speed_frac : svi->canal_speed_frac;
 		if (speed_frac > 0) c += YAPF_TILE_LENGTH * (1 + tf->tiles_skipped) * speed_frac / (256 - speed_frac);
 
+		/* Lock penalty. */
+		if (IsTileType(n.GetTile(), MP_WATER) && IsLock(n.GetTile()) && GetLockPart(n.GetTile()) == LOCK_PART_MIDDLE) {
+			const uint canal_speed = svi->ApplyWaterClassSpeedFrac(svi->max_speed, false);
+			c += (YAPF_TILE_LENGTH * canal_speed) / 16;
```

Comment:
```suggestion
			/* Cost is proportional to the vehicle's speed as the vehicle stops in the lock. */
			c += (4 * YAPF_TILE_LENGTH * canal_speed) / TILE_HEIGHT / 8;
```
Practically the 8 should be replaced with a constant associated with the 7 of `ShipMoveUpDownOnLock`. After all, the ship goes up 8 levels and waits 8 ticks between moves. In total that's 64.

To make it equivalent, that means the fudge factor would become 4 instead of 16, right? And there are many numeric YAPF_TILE_LENGTH fudge factors in YAPF, so that feels better than fudging the division here.